dev/src/bulk-writer.ts

```diff
@@ -211,25 +221,70 @@ class BulkCommitBatch {
     // Capture the error stack to preserve stack tracing across async calls.
     const stack = Error().stack!;
 
-    return this.writeBatch.bulkCommit().catch(err => {
-      throw wrapError(err, stack);
-    });
+    let results: BatchWriteResult[] = [];
+    for (let attempt = 0; attempt < MAX_RETRY_ATTEMPTS; attempt++) {
+      await this.backoff.backoffAndWait();
+
+      try {
+        results = await this.writeBatch.bulkCommit();
+      } catch (err) {
+        // Map the failure to each individual write's result.
+        results = [...this.pendingOps.keys()].map(key => {
+          return {key, writeTime: null, status: wrapError(err, stack)};
+        });
+      }
+      this.processResults(results, /* shouldRetry= */ true);
+
+      if (this.pendingOps.size > 0) {
+        logger(
+          'BulkWriter.bulkCommit',
+          null,
+          `Current batch failed at retry #${attempt}. Num failures: ` +
+            `${this.pendingOps.size}.`
+        );
+
+        this.writeBatch = new WriteBatch(this.firestore, this.writeBatch, [
+          ...this.pendingOps.keys(),
+        ]);
+      } else {
+        this.completedDeferred.resolve();
+        return;
```

Comment:
This could just be a break,  this.processResult() would be a no-op.

Reply:
The map entry is deleted from `pendingOps` after the promise is resolved in `processResults`, so an error would be thrown if `processResults` is called again.